### PR TITLE
ra: add support for direct array access notation 

### LINKED ISSUE #9958: Record Accessor does not support direct array access notation
## Bug Report

The Record Accessor API currently cannot handle direct array access patterns like $[0] or $array[0]. It requires the array to be nested within a map first, like $data.array[0].

```yaml
# This doesn't work
pipeline:
  inputs:
    - name: dummy
      dummy: '{"array": ["a", "b", "c"]}'
  outputs:
    - name: stdout
      match: "*"
      processors:
        logs:
          - name: calyptia
            actions:
              - type: put
                opts:
                  key: "worked"
                  value: "true"
                condition:
                  operator: AND
                  rules:
                    - field: "$array[0]"  # Returns empty
                      operator: eq
                      value: "a"

# This works
pipeline:
  inputs:
    - name: dummy
      dummy: '{"data": {"array": ["a", "b", "c"]}}'
  outputs:
    - name: stdout
      match: "*"
      processors:
        logs:
          - name: calyptia
            actions:
              - type: put
                opts:
                  key: "worked"
                  value: "true"
                condition:
                  operator: AND
                  rules:
                    - field: "$data.array[0]"  # Works correctly
                      operator: eq
                      value: "a"
```

The record accessor implementation in flb_ra_key_to_value requires a map lookup first through ra_key_val_id before handling any array access. This makes it impossible to access array elements directly without first having a map key.

### Proposed Solution:
Modify the record accessor code to:

1. Check if the root object is an array when direct array access is attempted
2. Add handling for array access without requiring an initial map lookup
3. Parse and validate array indices directly

### PR DESCRIPTION
**Pull request by niedbalski**
# Description
This PR fixes an issue where the Record Accessor API couldn't handle direct array access patterns like `$array[0]`.

The fix changes the condition that checks if a lookup succeeded by replacing `!found` with `matched == 0`, which properly tracks array element access.

Added test cases to verify array access works:
- Direct array access (`$array[0]`)
- Nested array access (`$matrix[1][2]`) 
- Mixed array and map access (`$records[1]['name']`)
- Handling of edge cases (out-of-bounds indices, nonexistent keys, type mismatches)

Fixes #9958

## Testing
- [x] Example configuration file for the change
```
pipeline:
  inputs:
    - name: dummy
      dummy: '{"array": ["a", "b", "c"]}'
  outputs:
    - name: stdout
      match: "*"
      processors:
        logs:
          - name: calyptia
            actions:
              - type: put
                opts:
                  key: "worked"
                  value: "true"
                condition:
                  operator: AND
                  rules:
                    - field: "$array[1]"
                      operator: eq
                      value: "b"
```

- [x] Debug log output from testing the change
```
[2025/02/25 14:55:32] [ info] [processor:calyptia:logs.0] added record accessor rule "$array[1]" eq "b"
[2025/02/25 14:55:32] [ info] [processor:calyptia:logs.0] rule condition result: true
[2025/02/25 14:55:32] [ info] [output:stdout:stdout.0] {"array":["a","b","c"],"worked":"true"}
```

- [x] Attached Valgrind output that shows no leaks or memory corruption was found
[N/A]

- [x] Run local packaging test showing all targets build
[N/A]

## Documentation
- [x] Documentation required for this feature
[N/A]

## Backporting
- [x] Backport to latest stable release.
[N/A]

### CODE REVIEW

**Review by pwhelan — APPROVED** [review]
LGTM

Must applaud the rigorous testing. Good work!

File: src/flb_ra_key.c [hunk 1]
```diff
@@ -150,40 +154,41 @@ static int subkey_to_object(msgpack_object *map, struct mk_list *subkeys,
             }
 
             /* Index limit and ensure no overflow */
-            if (entry->array_id == INT_MAX ||
-                cur.via.array.size < entry->array_id + 1) {
+            if (entry->array_id >= cur.via.array.size) {
```
Comment (edsiper) [inline src/flb_ra_key.c:157]: I think we still need to check the INT_MAX limit and array_id + 1 to prevent an integer overflow

Comment (edsiper) [reply src/flb_ra_key.c:157]: hmm maybe something like

```c
if (entry->array_id == INT_MAX || entry->array_id >= cur.via.array.size) {
    return -1;
}
```

File: src/flb_ra_key.c [hunk 2]
```diff
@@ -150,40 +154,41 @@ static int subkey_to_object(msgpack_object *map, struct mk_list *subkeys,
...
+
+            if (levels == matched) {
+                break;
+            }
+
+            continue;
         }
 
+        /* Handle map objects */
         if (cur.type != MSGPACK_OBJECT_MAP) {
```
Comment (edsiper) [inline src/flb_ra_key.c:174]: in this case, what will happen if the accessed value is another array ?

**Review by edsiper — CHANGES_REQUESTED** [review]

File: src/flb_ra_key.c (hunk 2 above)
Comment (niedbalski) [reply src/flb_ra_key.c:174]: Good question. When the accessed value is another array, the function will process it in the next iteration through if (entry->type == FLB_RA_KEY_ARRAY), which runs before the map type check. This is validated in cb_nested_array_access() test, which successfully accesses $matrix[1][2] from {"matrix": [[1, 2, 3], [4, 5, 6], [7, 8, 9]]} to get the value 6.

**Comment by niedbalski** [comment]
Please hold on merging here @edsiper , I think I found another condition to fix.
